build-system/tasks/performance/measure-documents.js

```diff
@@ -147,33 +299,43 @@ function writeMetrics(url, version, metrics) {
  *
  * @param {string} url
  * @param {string} version "control" or "experiment"
- * @param {Object} options
+ * @param {!Object} config
  * @return {Promise}
  */
-async function measureDocument(url, version, {headless}) {
+async function measureDocument(url, version, config) {
   const browser = await puppeteer.launch({
-    headless,
+    headless: config.headless,
     args: [
       '--allow-file-access-from-files',
       '--enable-blink-features=LayoutInstabilityAPI',
+      '--disable-web-security',
     ],
   });
 
   const page = await browser.newPage();
+  const handlerOptionsForUrl = {...config.urlToHandlers[url]};
+  const handlersList = [];
   await page.setCacheEnabled(false);
+  await page.setRequestInterception(true);
   await setupMeasurement(page);
+  setUpDefaultHandlers(handlersList);
+  setUpAddionalHandlers(handlersList, handlerOptionsForUrl);
+  startRequestListener(handlersList, page);
 
   try {
     await page.goto(`file:${urlToCachePath(url, version)}`, {
       waitUntil: 'networkidle0',
     });
-  } catch {
+  } catch (e) {
     // site did not load
     await browser.close();
+    console.log(e);
     return;
   }
 
   const metrics = await readMetrics(page);
+  await delayBasedOnHandlerOptions(handlerOptionsForUrl);
```

Comment:
If a request arrives right after you start the wait here, you would still need to wait for the entire timeout period. Please make this a promise race between request arrives vs timeout  with a defined startpoint.

Reply:
Will do